### PR TITLE
Implement WASM bulk memory operation and reference types proposals

### PR DESCRIPTION
**Pull request by q82419**
1. Add data count section.
2. New definition of `numtype` and `reftype`.
3. Update error code definitions.
4. Apply new version of `data segment` and `element segment`.
5. Add new bulk memory operation and reference type instructions.
  a. Reference instructions:
    * ref.null
    * ref.is_null
    * ref.func
  b. Table instructions:
    * table.get
    * table.set
    * table.init
    * elem.drop
    * table.copy
    * table.grow
    * table.size
    * table.fill
  c. Memory instructions:
    * memory.init
    * data.drop
    * memory.copy
    * memory.fill
  d. Parametric instructions:
    * select t
6. Update validator for data and element contexts.
7. Refactor runtime structure for supporting data and element instances.
8. Use WASM spec test suite for reference types.
9. Add document about using `externref`.

### CODE REVIEW

File: lib/ast/instruction.cpp [hunk 1]
```diff
@@ -290,31 +425,70 @@ Expect<void> MemoryInstruction::loadBinary(FileMgr &Mgr) {
...
-    LOG(ERROR) << ErrInfo::InfoAST(ASTNodeAttr::Instruction);
-    return Unexpect(Res);
+  switch (Code) {
+  /// Read the 0x00 checking code in memory.grow, memory.size, memory.fill, and
+  /// memory.copy cases.
+  case OpCode::Memory__copy:
+    if (auto Res = readCheck(); !Res) {
+      return Unexpect(Res);
+    }
+    // fall through
```
Comment (hydai) [inline lib/ast/instruction.cpp:438]: `[[fallthrough]];`

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/ast/instruction.cpp (hunk 1 above)
Comment (q82419) [reply lib/ast/instruction.cpp]: Done.

**Review by hydai — CHANGES_REQUESTED** [review]

File: include/runtime/instance/table.h [hunk 2]
```diff
@@ -28,64 +29,128 @@ namespace Instance {
...
   }
 
   /// Get boundary index.
   uint32_t getBoundIdx() const noexcept {
-    return ((MinSize > 0) ? (MinSize - 1) : 0);
+    return ((Refs.size() > 0) ? (Refs.size() - 1) : 0);
+  }
+
+  /// Grow table
+  bool growTable(const uint32_t Count, const ValVariant Val) {
```
[the remark this answers is not in the document]
Comment (q82419) [reply include/runtime/instance/table.h:65]: Done.

File: Changelog.md [hunk 3]
```diff
@@ -1,3 +1,65 @@
...
+    * call_indirect
+    * select
+  * Applied the new definition of `data count section`, `data segment`, and `element segment`.
+  * Applied validation for `data segment` and `element segment`.
+  * Added the `data instance` and `element instance`.
+  * Applied the new instantiation flow.
+
+Refactor:
+
+* Completed the enumeration value checking in loading phase.
```
Comment (hydai) [inline Changelog.md:40]: in `the` loading phase.

File: Changelog.md [hunk 4]
```diff
@@ -1,3 +1,65 @@
...
+* Updated the value type definition.
+  * `ValType` is updated to include `NumType` and `RefType`.
+  * `NumType` is updated to include `i32`, `i64`, `f32`, and `f64`.
+  * `RefType` is updated to include `funcref` and `externref`, which replaced the `ElemType`.
+* Updated error codes according to test suite for the reference types proposal.
```
Comment (hydai) [inline Changelog.md:45]: to `the` test suite...

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by hydai — APPROVED** [review]
LGTM

**Pull request merged**